File: BOJ/14502.cpp
#include <bits/stdc++.h>

using namespace std;
typedef pair<int, int> pii;

int n, m;

vector<pii> virus;
vector<pii> wall_cand;
int test_with_three_walls(vector<pii> walls, vector<vector<int>> arr) {
  vector<vector<int>> test_arr(arr);

  for (auto wall : walls) {
    test_arr[wall.first][wall.second] = 1;
  }

  // initial Queue
  queue<pii> q;
  for (auto virus : virus) {
    q.push(virus);
  }

  vector<pii> directions = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
  vector<vector<bool>> visited(n, vector<bool>(m, false));

  while (!q.empty()) {
    pii cur = q.front();
    q.pop();

    test_arr[cur.first][cur.second] = 2;
    visited[cur.first][cur.second] = true;

    for (auto dir : directions) {
      int nx = cur.first + dir.first;
      int ny = cur.second + dir.second;

      if (nx < 0 || nx >= n || ny < 0 || ny >= m) {
        continue;
      }

      if (test_arr[nx][ny] == 0 && !visited[nx][ny]) {
        visited[nx][ny] = true;
        q.push({nx, ny});
      }
    }
  }

  int count = 0;
  for (int i = 0; i < n; ++i) {
    for (int j = 0; j < m; ++j) {
      if (test_arr[i][j] == 0) {
        count++;
      }
    }
  }
  return count;
}

int main() {
  cin >> n >> m;

  vector<vector<int>> map(n, vector<int>(m, 0));

  for (int i = 0; i < n; ++i) {
    for (int j = 0; j < m; ++j) {
      cin >> map[i][j];
      if (map[i][j] == 0) {
        wall_cand.push_back({i, j});
      } else if (map[i][j] == 2) {
        virus.push_back({i, j});
      }
    }
  }
  int answer = 0;
  for (int i = 0; i < wall_cand.size(); i++) {
    for (int j = i + 1; j < wall_cand.size(); j++) {
      for (int k = j + 1; k < wall_cand.size(); k++) {
        vector<pii> walls = {wall_cand[i], wall_cand[j], wall_cand[k]};
        answer = max(answer, test_with_three_walls(walls, map));
      }
    }
  }

  cout << answer;

  return 0;
}

File: BOJ/14502.py
#first line
N, M = map(int, input().split())

arr = []
virus = []
wall_cand = set()


for i in range(N):
    arr.append(list(map(int, input().split())))
    for j in range(M):
        if arr[i][j] == 2:
            virus.append((i, j))
        elif arr[i][j] == 0:
            wall_cand.add((i, j))

from itertools import combinations
from collections import deque
from copy import deepcopy

def test_with_three_walls(walls, original_arr):
    test_arr = original_arr

    for wall_x, wall_y in walls:
        test_arr[wall_x][wall_y] = 1
    
    
    # bfs
    q = deque()
    q.extend(virus)
    dx = [0, 0, 1, -1]
    dy = [1, -1, 0, 0]
    visited = [[False] * M for _ in range(N)]
    
    while q:
        x, y = q.popleft()
        
        test_arr[x][y] = 2
        visited[x][y] = True

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0 <= nx < N and 0 <= ny < M and test_arr[nx][ny] == 0 and not visited[nx][ny]:
                q.append((nx, ny))
    
    # count
    count = 0
    for i in range(N):
        for j in range(M):
            if test_arr[i][j] == 0:
                count += 1


    for wall_x, wall_y in walls:
        test_arr[wall_x][wall_y] = 0
    
    return count
        
    
        
        

max_value = 0
for a,b,c in combinations(wall_cand, 3):
    max_value = max(max_value, test_with_three_walls((a,b,c), arr))
    

print(max_value)




